ngraph/core/include/openvino/core/type.hpp

```diff
@@ -108,14 +108,24 @@ as_type(Value value) {
     return ov::is_type<Type>(value) ? static_cast<Type*>(value) : nullptr;
 }
 
+template <typename T>
+struct AsTypePtr;
+
 /// Casts a std::shared_ptr<Value> to a std::shared_ptr<Type> if it is of type
 /// Type, nullptr otherwise
-template <typename Type, typename Value>
-typename std::enable_if<
-    std::is_convertible<decltype(std::static_pointer_cast<Type>(std::declval<Value>())), std::shared_ptr<Type>>::value,
-    std::shared_ptr<Type>>::type
-as_type_ptr(Value value) {
-    return ov::is_type<Type>(value) ? std::static_pointer_cast<Type>(value) : std::shared_ptr<Type>();
+template <typename In>
+struct AsTypePtr<std::shared_ptr<In>> {
+    template <typename Type>
+    static std::shared_ptr<Type> call(const std::shared_ptr<In>& value) {
+        return ov::is_type<Type>(value) ? std::static_pointer_cast<Type>(value) : std::shared_ptr<Type>();
+    }
+};
```

Comment:
can we move to details?

Reply:
it's not end user API, so want to hide under explicit name like `details`, `internal`

Reply:
We already have an internal `ov::util` namespace.  Let use it.

Reply:
ok